### PR TITLE
[CLI-1757] only get service accounts and users once, in kafka acl commands

### PR DESCRIPTION
**Pull request by MuweiHe**
Release Notes
-------------

Breaking Changes
- <PLACEHOLDER>

New Features
- <PLACEHOLDER>

Bug Fixes
- <PLACEHOLDER>

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----

only get service accounts and users once, in kafka acl commands, if users are already got, pass it on to next call.

References
----------

Test & Review
-------------

```
mhe@C02DV0KVMD6T cli % confluent kafka acl list -vvv
2023-02-28T13:20:38.937-0800 [DEBUG] Searching `$PATH` for plugins. Plugins can be disabled in /Users/mhe/.confluent/config.json.

2023-02-28T13:20:39.307-0800 [DEBUG] UserService.GetServiceAccounts request: GET https://confluent.cloud/api/service_accounts
2023-02-28T13:20:39.613-0800 [DEBUG] UserService.GetServiceAccounts response: 200 OK X-Request-Id:60382d3c0de95e8b100a1f8eb6f664e4 Body: {"users":[{"id":618212,"email":"57547-sa-2@serviceaccounts.confluent.cloud","first_name":"","last_name":"","deactivated":false,"verified":"1970-01-01T00:00:00Z","created":"2022-04-19T20:49:50.655621Z","modified":"2022-04-19T20:49:50.655621Z","service_name":"sa-2","service_description":"hello","service_account":true,"preferences":{},"internal":false,"resource_id":"sa-nv87wz","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_UNKNOWN"},{"id":1031339,"email":"57547-ksql.lksqlc-j33wym@serviceaccounts.confluent.cloud","first_name":"","last_name":"","deactivated":false,"verified":"1970-01-01T00:00:00Z","created":"2022-12-06T21:28:37.587147Z","modified":"2022-12-06T21:28:37.587147Z","service_name":"KSQL.lksqlc-j33wym","service_description":"SA for KSQL w/ ID lksqlc-j33wym and Name tester2","service_account":true,"preferences":{},"internal":false,"resource_id":"sa-5m211z","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_UNKNOWN"},{"id":1115274,"email":"57547-test-account@serviceaccounts.confluent.cloud","first_name":"","last_name":"","deactivated":false,"verified":"1970-01-01T00:00:00Z","created":"2023-01-20T18:01:52.048401Z","modified":"2023-01-20T18:01:52.048401Z","service_name":"test-account","service_description":"sa","service_account":true,"preferences":{},"internal":false,"resource_id":"sa-zp5qx0","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_UNKNOWN"},{"id":1168710,"email":"57547-hello@serviceaccounts.confluent.cloud","first_name":"","last_name":"","deactivated":false,"verified":"1970-01-01T00:00:00Z","created":"2023-02-15T23:28:51.900318Z","modified":"2023-02-15T23:28:51.900318Z","service_name":"hello","service_description":"trying","service_account":true,"preferences":{},"internal":false,"resource_id":"sa-55w02z","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_UNKNOWN"},{"id":1170527,"email":"57547-demoserviceaccount@serviceaccounts.confluent.cloud","first_name":"","last_name":"","deactivated":false,"verified":"1970-01-01T00:00:00Z","created":"2023-02-16T17:37:48.120023Z","modified":"2023-02-16T17:37:48.120023Z","service_name":"DemoServiceAccount","service_description":"This is a demo service account.","service_account":true,"preferences":{},"internal":false,"resource_id":"sa-1n9ykj","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_UNKNOWN"}],"page_info":{"page_size":0,"page_token":""},"error":null} request: GET https://confluent.cloud/api/service_accounts
2023-02-28T13:20:39.614-0800 [DEBUG] UserService.List request: GET https://confluent.cloud/api/users
2023-02-28T13:20:39.684-0800 [DEBUG] UserService.List response: 200 OK X-Request-Id:e4e95775db2e0507b09bf3b17b8d5bbf Body: {"users":[{"id":222045,"email":"mhe@confluent.io","first_name":"Muwei","last_name":"He","deactivated":false,"verified":"2021-04-28T20:42:34.233043Z","created":"2021-04-28T20:42:16.002626Z","modified":"2023-01-24T00:00:17.777761Z","service_name":"","service_description":"","service_account":false,"preferences":{"ContentfulAnnouncement_1bNdXQcWRrxk6NTXGv1ZwX":"COMPLETE","ContentfulAnnouncement_5ZmlnMShtNKJhWihFLOohr":"COMPLETE","ContentfulAnnouncement_dYhQIQYjVr5c691IKq4Ez":"COMPLETE","accountsWithStreamGovernancePackageUpsellSeen":"{\"env-5z5w8\":true,\"env-2mdy1\":true}","branchPageShown":"COMPLETE","gettingStarted":"SKIPPED","gettingStarted_accountId":"env-5z5w8","gettingStarted_clusterId":"","gettingStarted_lastCompleteStep":"clusterCreate","guidedLearning":"CLOSED","hasDismissedGlobalAnnouncement-a1299b47dcf2337bbdd9ad27b85f42046f0fb1edc8529940ec68516bd33eb9b2":"true","hasFinishedWelcomeSurvey":"true","ksqlGettingStarted":"IN_PROGRESS","ksqlGettingStarted_accountId":"env-2mdy1","ksqlGettingStarted_clusterId":"","ksqlGettingStarted_ksqlClusterId":"lksqlc-dowpky","ksqlGettingStarted_lastCompleteStep":"ksqlCreate","ksqlGettingStarted_source":"ksql-cta","showWelcomePageStatus":"COMPLETE","streamDesignerFlow":"STARTED","userReportedKafkaFamiliarity":"aware","userReportedObjective":"learn","userReportedRole":"developer"},"internal":false,"resource_id":"u-e02dw5","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_LOCAL"},{"id":623926,"email":"mhe+apple@confluent.io","first_name":"Muwei","last_name":"Cinder","deactivated":false,"verified":"1970-01-01T00:00:00Z","created":"2022-04-22T21:37:47.055572Z","modified":"2022-04-22T21:40:56.566582Z","service_name":"","service_description":"","service_account":false,"preferences":{},"internal":false,"resource_id":"u-j5pzoq","deactivated_at":null,"social_connection":"","auth_type":"AUTH_TYPE_LOCAL"}],"error":null,"page_info":{"page_size":0,"page_token":""}} request: GET https://confluent.cloud/api/users
```

Open Questions / Follow-ups
---------------------------

### CODE REVIEW

File: internal/cmd/kafka/command_acl_create.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ func (c *aclCommand) create(cmd *cobra.Command, _ []string) error {
 		return err
 	}
 
-	userIdMap, err := c.mapResourceIdToUserId()
+	users, userIdMap, err := c.mapResourceIdToUserId(nil)
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_acl_create.go:58]: Would it make more sense to list the service accounts and users outside of this function, and pass in the combined list?

Comment (sgagniere) [reply internal/cmd/kafka/command_acl_create.go:58]: +1 on that suggestion

File: internal/cmd/kafka/command_acl.go [hunk 2]
```diff
@@ -140,48 +139,20 @@ func parsePrincipal(principal string) (string, error) {
...
-
-		adminUsers, err := c.Client.User.List(context.Background())
-		if err != nil {
-			return nil, nil, err
-		}
-
-		users = append(serviceAccounts, adminUsers...)
-	}
-
+func (c *aclCommand) mapUserIdToResourceId(users []*ccloud.User) (map[int32]string, error) {
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_acl.go:142]:
```suggestion
func (c *aclCommand) mapUserIdToResourceId(users []*ccloud.User) map[int32]string {
```

File: internal/cmd/kafka/command_acl_create.go [hunk 3]
```diff
@@ -55,7 +56,19 @@ func (c *aclCommand) create(cmd *cobra.Command, _ []string) error {
 		return err
 	}
 
-	users, userIdMap, err := c.mapResourceIdToUserId(nil)
+	serviceAccounts, err := c.Client.User.GetServiceAccounts(context.Background())
+	if err != nil {
+		return err
+	}
+
+	adminUsers, err := c.Client.User.List(context.Background())
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_acl_create.go:64]: Are we sure that this function only returns admin users? I thought it returned all users.

File: internal/cmd/kafka/command_acl_delete.go [hunk 4]
```diff
@@ -49,7 +50,19 @@ func (c *aclCommand) delete(cmd *cobra.Command, _ []string) error {
...
+	if err != nil {
+		return err
+	}
+
+	adminUsers, err := c.Client.User.List(context.Background())
+	if err != nil {
+		return err
+	}
+
+	users := append(serviceAccounts, adminUsers...)
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_acl_delete.go:63]: This is repeated 3 times! Let's make a helper function and put it in the top-level command_acl.go file?

File: internal/cmd/kafka/command_acl_create.go (hunk 3 above)
Comment (MuweiHe) [reply internal/cmd/kafka/command_acl_create.go:64]: Pretty sure

**Review by brianstrauch — APPROVED** [review]
So clean! :soap: Hopefully we'll be able to get rid of the numeric ID to resource ID conversions soon.

**Pull request merged**